app/models/concerns/document.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+require 'active_support/concern'
+
+module Document
+  extend ActiveSupport::Concern
+
+  included do # rubocop:disable Metrics/BlockLength
+    has_many :user_documents, dependent: :nullify
+    has_many :users, -> { distinct }, through: :user_documents
+    has_many :shares, -> { merge(Share.kept) }, dependent: :nullify, inverse_of: :document, foreign_key: 'document_id'
+
+    def heading
+      to_s
+    end
+
+    def document_type
+      self.class.to_s
+    end
+
+    def add_to(user:)
+      user_documents.create(user: user)
```

Comment:
This might be premature optimisation since the only thing using `Document` is overwriting all of these methods, but it should make it painless to add it to a new class